### PR TITLE
added diagnostic(s) support for fault-tolerance Asynchronous annotation on type

### LINKED ISSUE #74: Diagnostics for mp fault-tolerance Asynchronous annotation
https://github.com/eclipse/microprofile-fault-tolerance/blob/master/spec/src/main/asciidoc/asynchronous.asciidoc#asynchronous

Any methods marked with this annotation must return one of:

`java.util.concurrent.Future`
`java.util.concurrent.CompletionStage`
Otherwise, org.eclipse.microprofile.faulttolerance.exceptions.FaultToleranceDefinitionException occurs during deployment.

### PR DESCRIPTION
**Pull request by AlexXuChen**
Added diagnostic(s) support for fault-tolerance Asynchronous annotation marked on type declaration.

Fixes #74

Signed-off-by: Alexander Chen <alchen@redhat.com>

### CODE REVIEW

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java [hunk 1]
```diff
@@ -103,11 +103,37 @@ public boolean visit(MethodDeclaration node) {
...
+			List modifiers = type.modifiers();
+			for (Object modifier : modifiers) {
+				if (modifier instanceof Annotation) {
+					Annotation annotation = (Annotation) modifier;
+					if (isMatchAnnotation(annotation, ASYNCHRONOUS_ANNOTATION)) {
+						try {
+							MethodDeclaration[] methods = type.getMethods();
+							for (MethodDeclaration node : methods) {
+								MarkerAnnotation markAnnotation = (MarkerAnnotation) modifier;
+								validateAsynchronousAnnotation(node, diagnostics, context, markAnnotation);
```
Comment (rgrunber) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java:125]: I would delete `markerAnnotation` and just pass `(MarketAnnotation) modifier` in as an argument

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java [hunk 2]
```diff
@@ -103,11 +103,37 @@ public boolean visit(MethodDeclaration node) {
 			try {
 				validateMethod(node, diagnostics, context);
 			} catch (JavaModelException e) {
-				LOGGER.log(Level.WARNING, "An exception occurred when attempting to validate the fallback method");
+				LOGGER.log(Level.WARNING, "An exception occurred when attempting to validate the annotation marked method");
 			}
 			super.visit(node);
 			return true;
 		}
+
```
Comment (rgrunber) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java:111]: Remove the extraneous tab(s) on this line.

**Review by rgrunber — COMMENTED** [review]
Looks good overall. Could we add some testcases for this also ?

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java [hunk 3]
```diff
@@ -185,7 +208,10 @@ private void validateAsynchronousAnnotation(MethodDeclaration node, List<Diagnos
...
 				throw e;
 			}
-			if ((!(FUTURE_TYPE_UTILITY.equals(methodReturnTypeString)) && !(COMPLETION_STAGE_TYPE_UTILITY.equals(methodReturnTypeString)))) {
+			if ((!(FUTURE_TYPE_UTILITY.equals(methodReturnTypeString)) && 
+					!(COMPLETION_STAGE_TYPE_UTILITY.equals(methodReturnTypeString)) &&
+					!(UNI_TYPE_UTILITY.equals(methodReturnTypeString))
+					)) {
 				IOpenable openable = context.getTypeRoot().getOpenable();
 				Diagnostic d = context.createDiagnostic(context.getUri(),
 						"The annotated method does not return an object of type Future or CompletionStage",
```
Comment (rgrunber) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java:217]: Might as well expand this to include `io.smallrye.mutiny.Uni` ? I guess we could just call it Uni for now.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java [hunk 4]
```diff
@@ -185,7 +208,10 @@ private void validateAsynchronousAnnotation(MethodDeclaration node, List<Diagnos
 			} catch (Exception e) {
 				throw e;
 			}
-			if ((!(FUTURE_TYPE_UTILITY.equals(methodReturnTypeString)) && !(COMPLETION_STAGE_TYPE_UTILITY.equals(methodReturnTypeString)))) {
+			if ((!(FUTURE_TYPE_UTILITY.equals(methodReturnTypeString)) &&
```
Comment (rgrunber) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java:211]: There's an extra whitespace at the end of this line.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/java/MicroProfileFaultToleranceJavaDiagnosticsTest.java [hunk 5]
```diff
@@ -88,6 +88,38 @@ public void asynchronousNonFutureOrCompletionStage() throws Exception {
...
+				.getFile(new Path("src/main/java/org/acme/AsynchronousFaultToleranceClassResource.java"));
+		diagnosticsParams.setUris(Arrays.asList(javaFile.getLocation().toFile().toURI().toString()));
+		diagnosticsParams.setDocumentFormat(DocumentFormat.Markdown);
+
+		Diagnostic d1 = d(32, 11, 17, 
+				"The annotated method does not return an object of type Future or CompletionStage", 
+				DiagnosticSeverity.Error,
+				MicroProfileFaultToleranceConstants.DIAGNOSTIC_SOURCE,
+				MicroProfileFaultToleranceErrorCode.FAULT_TOLERANCE_DEFINITION_EXCEPTION);
+
```
Comment (rgrunber) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/java/MicroProfileFaultToleranceJavaDiagnosticsTest.java:107]: There's some extraneous whitespace in this file. You can usually see these if you run `git show` on your commit, in the commandline.

There's also an option in Eclipse to show whitespace characters and remove extraneous ones.

**Review by rgrunber — COMMENTED** [review]
Looks good overall. Just some minor things.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java (hunk 3 above)
Comment (AlexXuChen) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java:217]: Correct me if I'm wrong, I think UNI_TYPE_UTILITY on line 213 covers `io.smallrye.mutiny.Uni`

Comment (rgrunber) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java]: Yes, but the error message, right here, makes no indication of this.

Comment (AlexXuChen) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java]: Ah good catch

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/java/MicroProfileFaultToleranceJavaDiagnosticsTest.java [hunk 6]
```diff
@@ -67,27 +67,53 @@ public void asynchronousNonFutureOrCompletionStage() throws Exception {
...
 				MicroProfileFaultToleranceErrorCode.FAULT_TOLERANCE_DEFINITION_EXCEPTION);
-		
+
+		Diagnostic d2 = d(36, 11, 15,
+				"The annotated method does not return an object of type Future or CompletionStage",
+				DiagnosticSeverity.Error, MicroProfileFaultToleranceConstants.DIAGNOSTIC_SOURCE,
+				MicroProfileFaultToleranceErrorCode.FAULT_TOLERANCE_DEFINITION_EXCEPTION);
+
+		Diagnostic d3 = d(40, 11, 36,
+				"The annotated method does not return an object of type Future or CompletionStage",
```
Comment (rgrunber) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/java/MicroProfileFaultToleranceJavaDiagnosticsTest.java:110]: I think you just need t update these with Uni and it should be good.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/java/MicroProfileFaultToleranceJavaDiagnosticsTest.java [hunk 7]
```diff
@@ -67,27 +67,53 @@ public void asynchronousNonFutureOrCompletionStage() throws Exception {
...
+		IJDTUtils utils = JDT_UTILS;
+
+		MicroProfileJavaDiagnosticsParams diagnosticsParams = new MicroProfileJavaDiagnosticsParams();
+		IFile javaFile = javaProject.getProject()
+				.getFile(new Path("src/main/java/org/acme/AsynchronousFaultToleranceClassResource.java"));
+		diagnosticsParams.setUris(Arrays.asList(javaFile.getLocation().toFile().toURI().toString()));
+		diagnosticsParams.setDocumentFormat(DocumentFormat.Markdown);
+
+		Diagnostic d1 = d(32, 11, 17,
+				"The annotated method does not return an object of type Future or CompletionStage or Uni",
```
Comment (rgrunber) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/java/MicroProfileFaultToleranceJavaDiagnosticsTest.java:100]: This should fail right ? There's a comma right after Future in the original text.

Comment (AlexXuChen) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/java/MicroProfileFaultToleranceJavaDiagnosticsTest.java:100]: Ah yes sorry

Comment (rgrunber) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/java/MicroProfileFaultToleranceJavaDiagnosticsTest.java:100]: Please make sure you run the tests locally, particularly if modifying test cases and functionality like this.

Comment (AlexXuChen) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/java/MicroProfileFaultToleranceJavaDiagnosticsTest.java:100]: Forgot a save

**Pull request merged**